### ISSUE TITLE
referencing partials within coffeekup views

### ISSUE THREAD

**Issue by flippyhead** [mauricemach/coffeekup#11]
How do I include partials within coffeekup views?

**Comment by flippyhead** [comment]
I tried referring to just "partial" but get the error: 

ReferenceError: self is not defined

at eval at <anonymous> (/usr/local/lib/node/.npm/coffeekup/0.2.0/package/lib/coffeekup.js:186:12)
    at Object.<anonymous> (eval at <anonymous> (/usr/local/lib/node/.npm/coffeekup/0.2.0/package/lib/coffeekup.js:186:12))
    at anonymous (eval at <anonymous> (/usr/local/lib/node/.npm/coffeekup/0.2.0/package/lib/coffeekup.js:186:12))
    at Object.render (/usr/local/lib/node/.npm/coffeekup/0.2.0/package/lib/coffeekup.js:206:12)
    at ServerResponse.render (/usr/local/lib/node/.npm/express/1.0.0/package/lib/express/view.js:333:22)

**Comment by mauricemach** [comment]
Partials are not a feature in CoffeeKup itself, and that's usually the case with templating engines. They're often implemented at the web framework level, as I just did with zappa. I'm seriously considering offering this functionality (along with layouts) at the templating engine level though, especially since it can also be used from the client-side.

Meanwhile, you could define a `partial` function that would forward parameters to `coffeekup.render`, and put it in locals.

**Comment by focusaurus** [comment]
Any updates on this?  Can someone point me to a working example of using express partials with coffeekup templates?  I'm getting a similar error to flippyhead.

**Comment by emmalemma** [comment]
As near as I can figure, it was a silly coincidence that CoffeeKup ever _seemed_ to work with Express at all... (Edit: I now see it's because of the Express 2.0 changes. It's been a silly coincidence that it's seemed to slightly work since then :) After poking at it for a bit, I finally managed to get layouts working with this:

```
app.register '.coffee',
    compile:(text, options)->
        template = coffeekup.compile text
        (context)->
            template(context: context)
```

Which worked, and got partials working as context variables (since Express sends partial as one). But partials wouldn't output directly, so I had to write something like
    text @partial 'login'
which is just silly, since helpers should be in local scope. So I kept poking.

Turns out the reason for the "self is not defined" error is that CoffeeKup turns local variables into strings which are eval'd. Good for performance, bad if you need to refer to an external view in Express. You can change this behavior by passing {dynamic_locals: true} into coffeekup.compile, although that has a bit of a performance impact— 165ms vs 140ms in cake benchmark for me. Even so, I almost stopped there.

But then I found an elegant if hacky solution, since eval'd locals can still access the context scope, which comes with partial. So just make that line something like

```
app.register '.coffee',
    compile:(text, options)->
        template = coffeekup.compile text,
            locals: 
                partial: (args...)->text(ck_options.context.partial args...)
        (context)->
            template(context: context)
```

et voila! Looks pretty ugly, and I'm sure there's a prettier way to make this work, but it does. Now
    partial 'login'
does just what you expect, it's 05:30 here, and I'm going to bed :)

**Comment by mauricemach** [comment]
cushman, thanks for your detailing your struggle with coffeekup and solution. The new [0.2.4beta](https://github.com/mauricemach/coffeekup/tree/0.2.4beta) removes the need for your first solution, since it puts everything it receives at the "root" of the object param at the "context". So you'll be able to `text @partial 'login'` by default.

To transform this into `partial 'login'`, I think I'm going to add an express adapter to coffeekup, in the same way I already did with meryl. This adapter would work much like your second solution, creating proxy locals to certain utilities express sends us at the root, and writing to the buffer automatically when appropriate. So we'd only have to write:

```
app.register '.coffee', require('coffeekup').adapters.express
```

And be set, without grafting express-specific code into CK itself.

**Comment by mauricemach** [comment]
Landed in master (0.2.4beta was merged into master and will be published to npm as 0.3.0beta).
